From 5c6b96f0f862eca0cef97fe33bfd00b8685ad275 Mon Sep 17 00:00:00 2001
From: Thomas Langer <thomas.langer@intel.com>
Date: Wed, 6 Nov 2019 16:19:28 +0100
Subject: [PATCH] PONRTSYS-5112: Allow relocation for kernel on bootcore

This makes the bootcore build with CONFIG_SYS_SUPPORTS_RELOCATABLE
and still finding the appended dtb.
---
 arch/mips/lantiq/Kconfig                |  2 ++
 arch/mips/lantiq/grx500_bootcore/prom.c | 19 +++++++++++--------
 2 files changed, 13 insertions(+), 8 deletions(-)

diff --git a/arch/mips/lantiq/Kconfig b/arch/mips/lantiq/Kconfig
index 4ec7f8cd36d8..bc7ca091db41 100755
--- a/arch/mips/lantiq/Kconfig
+++ b/arch/mips/lantiq/Kconfig
@@ -62,10 +62,12 @@ config SOC_GRX500
 config SOC_GRX500_BOOTCORE
 	bool "GRX500_BOOTCORE"
 	select SOC_TYPE_GRX500_TEP
+	select SYS_SUPPORTS_RELOCATABLE
 
 config SOC_PRX300_BOOTCORE
 	bool "PRX300_BOOTCORE"
 	select SOC_TYPE_GRX500_TEP
+	select SYS_SUPPORTS_RELOCATABLE
 
 endchoice
 
diff --git a/arch/mips/lantiq/grx500_bootcore/prom.c b/arch/mips/lantiq/grx500_bootcore/prom.c
index 7010a6bf9919..112d407d1048 100644
--- a/arch/mips/lantiq/grx500_bootcore/prom.c
+++ b/arch/mips/lantiq/grx500_bootcore/prom.c
@@ -35,6 +35,7 @@
 /* access to the ebu needs to be locked between different drivers */
 DEFINE_SPINLOCK(ebu_lock);
 EXPORT_SYMBOL_GPL(ebu_lock);
+static __initdata const void *dtb;
 
 /*
  * this struct is filled by the soc specific detection code and holds
@@ -90,10 +91,9 @@ static void __init prom_init_cmdline(void)
 }
 //#endif
 
-static void plat_early_init_devtree(void)
-{
-	void *dtb = NULL;
 
+void __init *plat_get_fdt(void)
+{
 	/*
 	 * Load the builtin devicetree. This causes the chosen node to be
 	 * parsed resulting in our memory appearing
@@ -102,9 +102,10 @@ static void plat_early_init_devtree(void)
 		dtb = (void *)fw_passed_dtb;
 	else if (__dtb_start != __dtb_end)
 		dtb = (void *)__dtb_start;
+	else
+		panic("no dtb found");
 
-	if (dtb)
-		__dt_setup_arch(dtb);
+	return (void *)dtb;
 }
 
 void __init plat_mem_setup(void)
@@ -116,7 +117,8 @@ void __init plat_mem_setup(void)
 
 	set_io_port_base((unsigned long) KSEG1);
 
-	plat_early_init_devtree();
+	if (dtb)
+		__dt_setup_arch((void *)dtb);
 }
 
 void __init device_tree_init(void)
@@ -135,9 +137,10 @@ void __init prom_init(void)
 		soc_info.name, soc_info.rev_type);
 	soc_info.sys_type[LTQ_SYS_TYPE_LEN - 1] = '\0';
 	pr_info("SoC: %s\n", soc_info.sys_type);
-//#ifndef CONFIG_USE_EMULATOR
+
 	prom_init_cmdline();
-//#endif
+	plat_get_fdt();
+	BUG_ON(!dtb);
 }
 
 void __init ltq_soc_detect(struct ltq_soc_info *i)
